text/0071-trailing-commas.md

```diff
@@ -0,0 +1,187 @@
+objects, without adjusting commas.
+
+Many languages allow for trailing commas, including:
+* Rust
+* Python
+* Java,
+* Go
+* JavaScript
+* OCaml
+
+(Frustratingly: not JSON!)
+
+In addition, it will make our schema grammar and policy grammar more
+consistent, as the schema grammar allows trailing commas in many places.
+
+## Detailed design
+
+### Cedar Policy Grammar
+
+We want to allow trailing commas in the following expressions:
+1. scopes: `permit(principal, action, resource, )` 
+2. sets: `[1,2,3,]`
+3. records: `{ foo : 1, bar : 3, }`
+4. all method/function calls: `.contains(3,)`, `ip("10.10.10.10",)`
+
+In the grammar:
+
+The following grammar rules change:
+```
+Scope ::= Principal ',' Action ',' Resource
+```
+becomes:
+```
+Scope ::= Principal ',' Action ',' Resource ','?
+```
+
+```
+EntList ::= Entity {',' Entity}
+```
+becomes:
+```
+EntList ::= Entity {',' | ',' Entity} ','?
+```
+
+```
+ExprList ::= Expr {',' Expr}
+```
+becomes:
+```
+ExprList ::= Expr {',' | ',' Expr} ','?
+```
+
+```
+RecInits ::= (IDENT | STR) ':' Expr {',' (IDENT | STR) ':' Expr}
+```
+becomes:
+```
+RecInits ::= (IDENT | STR) ':' Expr {',' | ',' (IDENT | STR) ':' Expr}
+```
+
+### Cedar Schema Grammar
+The schema grammar already allows trailing commas in several places
+(such as record types and entity lists), but not everywhere:
+
+We want to allow trailing commas in the following expressions:
+1. membership lists: `entity Photo in [Account, Album,];`
+2. Multi-entity declarations: `entity Photo,Video;`
+3. Record types: `{ foo : Long, bar : String, }` (currently allowed)
+4. Multi-action declarations: `action foo,bar,;`
+5. Apply specs: `appliesTo { principal : ["A"], resource: ["B"],"`
+   (currently allowed)
+
+
+```
+EntTypes  := Path {',' Path}
+```
+becomes:
+
+```
+EntTypes  := Path {',' | ',' Path}
+```
+
+```
+Names     := Name {',' Name}
+```
+becomes:
+```
+Names     := Name {',' | ',' Name} ','?
+```
+
+```
+Idents    := IDENT {',' IDENT}
+```
+becomes:
+```
+Idents    := IDENT {',' | ',' IDENT}
+```
+
+### Cedar Policy Formatter
+This RFC proposes a change to the style guide, that having a trailing
+comma is the preferred style _except_ for the commas separating
+`principal`, `action`, and `resource`.
+The Cedar policy formatter should enforce this.
+
+Example:
+```
+permit(principal,action,resource,) when {
+    { 
+        "foo" : [1,2],
+        "bar" : 3
+    } == context
+}
+```
+would become:
+```
+permit(principal,action,resource) when {
+    { 
+        "foo" : [1,2,],
+        "bar" : 3,
+    } == context
+}
+```
+
+### Backwards Compatibility 
+This change is fully backwards compatible. 
+
+## Drawbacks
+
+* Trailing commas are potentially confusing.
+* Policies produced by the new formatter won't parse with old parsers
+
+## Alternatives
+
+* Keep status quo: this is a non critical change
+
+## Unresolved questions
+
+* We could forbid the trailing the comma after `resource`.
```

Comment:
I'm in favor of allowing it, particularly since scopes are so often split across multiple lines

Reply:
I'm also in favor of allowing it. No need to add an extra special case to worry about.